File: 백준/Gold/2294. 동전 2/동전 2.java
import java.io.*;
import java.util.*;

public class Main {

    private static int n;
    private static int k;
    private static int[] coins;
    private static int[] dp;

    public static void main(String[] args) throws IOException {
        BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer tokenizer = new StringTokenizer(input.readLine());
        n = Integer.parseInt(tokenizer.nextToken());
        k = Integer.parseInt(tokenizer.nextToken());
        coins = new int[n];
        dp = new int[k + 1];
        for (int i = 0; i < coins.length; i++) {
            coins[i] = Integer.parseInt(input.readLine());
        }
        Arrays.fill(dp, Integer.MAX_VALUE);
        dp[0] = 0;
        for(int coin : coins) {
            for (int i = coin; i <= k; i++) {
                if (dp[i - coin] != Integer.MAX_VALUE) {
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }
        System.out.println(dp[k] != Integer.MAX_VALUE ? dp[k] : -1);
    }
}

File: 백준/Gold/2294. 동전 2/동전 2.cc
#include <bits/stdc++.h>
using namespace std;
int n, k;
int coin[100];
int money[10001];
const int INF = 98765421;
int main(){
    cin >> n >> k;
    for (int i = 0; i < n; i++) {
        cin >> coin[i];
    }

    fill(money, money + 10001, INF);
    money[0] = 0;
    for (int i = 0; i <= 10000; i++) {
        for (int j = 0; j < n; j++) {
            if (i + coin[j] > 10000) continue;
            money[i + coin[j]] = min(money[i + coin[j]], money[i] + 1);
        }
    }

    if (money[k] == INF) {
        cout << -1 << '\n';
        return 0;
    } 
    
    cout << money[k] << '\n';
    return 0;
}